### PR TITLE
feat(ingestion): allow multiple metabase databases with same platform

### PR DESCRIPTION
**Pull request by remisalmon**
This PR allows configuring multiple databases of the same platform (postgres, snowflake, ...) in Metabase, currently not possible.

For example, this allows to map 2 Postgres databases `abc` and `xzy` to `db1` and `db2` in DataHub by using:
```yaml
     database_alias_map:
      postgres:
        "abc": "db1"
        "xyz": "db2"
```

Currently the config format is:
```yaml
     database_alias_map:
      postgres: "db"
```
which maps all postgres database to `db` in DataHub.

Tested locally with `DATAHUB_VERSION="v0.8.43"`, `acryl-datahub==0.8.43.1`.

## Checklist
- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [x] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [x] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

**Comment by remisalmon** [comment]
Added a commit to fix the lineage when Metabase queries are using fully qualified tables that may be referencing another database (also makes the code a little easier to read).

In Metabase it is possible to do, with a query on db1: `select table1.*, table2.* from schema1.table1 join db2.schema2.table2 on ...` where `table1` is in `db1.schema1` (the query's default db) and `table2` in `db2.schema2` (maybe not a best practice but still possible...).

Tested locally.

File: metadata-ingestion/src/datahub/ingestion/source/metabase.py [hunk 1]
```diff
@@ -441,7 +441,7 @@ def get_datasource_urn(self, card_details):
             card_details.get("database_id", "")
         )
         query_type = card_details.get("dataset_query", {}).get("type", {})
-        source_paths = set()
+        source_tables = set()
```
Comment (remisalmon) [inline metadata-ingestion/src/datahub/ingestion/source/metabase.py:444]: `source_paths` was a set of `schema.table` strings.
`source_tables` is a set of fully qualified `database.schema.table`  strings, moving some existing logic around to get the database name for "native" queries (`query_type == "query"`) and adding new logic to get the database name for SQL queries (`query_type != "query"`).

**Comment by remisalmon** [comment]
@rslanka I rebased on the latest master (there is 1 failing check but looks unrelated to this PR?), updated/fixed the documentation and left a comment in the code to help the review. I hope this helps, thanks!

**Comment by rslanka** [comment]
Thanks @remisalmon. I'll take a look soon :-)

File: metadata-ingestion/src/datahub/ingestion/source/metabase.py [hunk 2]
```diff
@@ -244,7 +244,9 @@ def construct_dashboard_from_api_data(
         chart_urns = []
         cards_data = dashboard_details.get("ordered_cards", "{}")
         for card_info in cards_data:
-            chart_urn = builder.make_chart_urn(self.platform, card_info.get("id", ""))
+            chart_urn = builder.make_chart_urn(
+                self.platform, card_info.get("card_id", "")
+            )
```
Comment (remisalmon) [inline metadata-ingestion/src/datahub/ingestion/source/metabase.py:249]: Not sure if this has changed in the Metabase API or if it has always been broken but `card_id` is the correct "foreign key" to `id` in the api/card enpoint.

**Comment by jjoyce0510** [comment]
Hi there. We are considering whether to accept this or not. Determining our stance on aliases across sources.

Thanks for your patience, 
John

**Comment by mayurinehate** [comment]
Hi @remisalmon This is indeed a problem that `platform_instance_map` in some sources(including metabase) do not allow setting different platform instances where tool(metabase) connects to multiple instances of same platform.

The solution you've proposed does not solve the case with different instances having same database.

Ideal would be to use `platform_instance_map` as a mapping between datasource identifier in the tool(metabase) and platform instance identifier in DataHub. What do you think about using the database identifier accepted in metabase REST API - https://www.metabase.com/docs/latest/api/database ?

**Comment by remisalmon** [comment]
> Hi @remisalmon This is indeed a problem that `platform_instance_map` in some sources(including metabase) do not allow setting different platform instances where tool(metabase) connects to multiple instances of same platform.
> 
> The solution you've proposed does not solve the case with different instances having same database.
> 
> Ideal would be to use `platform_instance_map` as a mapping between datasource identifier in the tool(metabase) and platform instance identifier in DataHub. What do you think about using the database identifier accepted in metabase REST API - https://www.metabase.com/docs/latest/api/database ?

I agree having multiple databases of the same platform with the same name will be an issue (but also a very edge case), using the database ID from the API should work assuming the user configuring the metabase source in DataHub has access to the content of `api/database`, I will try to push those changes when I have a chance.

**Comment by remisalmon** [comment]
> Hey @remisalmon how fair is the chance that user configuring metabase source has access to contents of database ID ? Does database id show up anywhere in metabase UI or accessible only through API ?

The database ID does show up in the URL when browsing the data from the UI as `https://my.metabase.instance.com/browse/2-my-database` where `2` is the database ID, at least with Metabase v0.43 but may not be very reliable... 

Otherwise admin access is required to get the API response, I think it is fair to assume that the user configuring the metabase source would have access to the API response either directly or through an admin (the API response does not contain passwords), plus `database_alias_map` is an optional field in the config.

(edit: I edited my orginal post I got confused between `database_alias_map` which this PR changes vs `platform_instance_map`)

**Comment by mayurinehate** [comment]
Oh, It is confusing indeed. Why should need both  `database_alias_map` and `platform_instance_map` ?  I think it should be sufficient to have only one config `platform_instance_map` that maps metabase database id with platform instance, unless I am missing something.

Is database name from metabase not same as database name from source platform, such as postgres ?

**Comment by remisalmon** [comment]
> Oh, It is confusing indeed. Why should need both `database_alias_map` and `platform_instance_map` ? I think it should be sufficient to have only one config `platform_instance_map` that maps metabase database id with platform instance, unless I am missing something.
> 
> Is database name from metabase not same as database name from source platform, such as postgres ?

`platform_instance_map` i unrelated to this PR and deals with https://datahubproject.io/docs/platform-instances/.

The database name from Metabase is a user-friendly name like "My Database"; the actual database may be something like "a1b2c3" and we use ` database_alias_map` to rename "a1b2c3" to "my_database" in DataHub.
A use case of this PR would be when the same database "a1b2c3" was ingested with the Postgres source with `database_alias` set to "my_database": https://datahubproject.io/docs/generated/ingestion/sources/postgres/).

**Comment by hsheth2** [comment]
@remisalmon In general we've been trying to move towards using platform instances and `platform_instance_map`, and are working on deprecating `database_alias` and similar fields. The reasoning is that platform instance services this need in a more first-class way throughout DataHub and doesn't require this sort of nested database alias mapping. In addition, we want to keep the generated URNs similar to the underlying platform.

I'm not sure how much you've already built using database_alias into your system, which is why I'm not 100% against merging this PR. However, this does feel like a very specific edge case and I would prefer that we use the already existing `platform_instance_map` option if possible instead of adding this.

**Comment by remisalmon** [comment]
> @remisalmon In general we've been trying to move towards using platform instances and `platform_instance_map`, and are working on deprecating `database_alias` and similar fields. The reasoning is that platform instance services this need in a more first-class way throughout DataHub and doesn't require this sort of nested database alias mapping. In addition, we want to keep the generated URNs similar to the underlying platform.
> 
> I'm not sure how much you've already built using database_alias into your system, which is why I'm not 100% against merging this PR. However, this does feel like a very specific edge case and I would prefer that we use the already existing `platform_instance_map` option if possible instead of adding this.

Hi @hsheth2, `platform_instance_map` does what I wanted to get to with this PR and I see `database_alias` is going to deprecated in the postgres source, I am closing this PR. Thanks all for the comments.
There are still issues with the metabase source with mixed upper/lower case identifiers and default schema names that could be picked up from the API, I will try to push another PR.

**Comment by hsheth2** [comment]
Sounds good, thanks @remisalmon.

**Pull request closed without merging**